kyo-prelude/shared/src/main/scala/kyo/Batch.scala

```diff
@@ -191,8 +186,9 @@ object Batch:
     object internal:
         type Source[A, B, S] = Seq[A] => (A => B < S) < S
 
-        type Op[V, -S] = Seq[V] | Call[?, V, S]
-        case class Call[A, B, -S](v: A, source: Source[A, B, S])
+        sealed trait Op[A]
+        case class Eval[A](seq: Seq[A])                         extends Op[A]
+        case class Call[A, B, S](v: A, source: Source[A, B, S]) extends Op[B < (Batch & S)]
```

Comment:
I couldn't keep the optimization to avoid allocations for `Batch.eval` since it's necessary to indicate from the `Op` that the result is a computation in the case of `Call`.